doc/reference/kernel/scheduling/index.rst

```diff
@@ -0,0 +1,260 @@
+.. _scheduling_v2:
+
+Scheduling
+##########
+
+The kernel's priority-based scheduler allows an application's threads
+to share the CPU.
+
+Concepts
+********
+
+The scheduler determines which thread is allowed to execute
+at any point in time; this thread is known as the **current thread**.
+
+Whenever the scheduler changes the identity of the current thread,
+or when execution of the current thread is supplanted by an ISR,
+the kernel first saves the current thread's CPU register values.
+These register values get restored when the thread later resumes execution.
+
+Thread States
+=============
+
+A thread that has no factors that prevent its execution is deemed
+to be **ready**, and is eligible to be selected as the current thread.
+
+A thread that has one or more factors that prevent its execution
+is deemed to be **unready**, and cannot be selected as the current thread.
+
+The following factors make a thread unready:
+
+* The thread has not been started.
+* The thread is waiting on for a kernel object to complete an operation.
+  (For example, the thread is taking a semaphore that is unavailable.)
+* The thread is waiting for a timeout to occur.
+* The thread has been suspended.
+* The thread has terminated or aborted.
+
+Thread Priorities
+=================
+
+A thread's priority is an integer value, and can be either negative or
+non-negative.
+Numerically lower priorities takes precedence over numerically higher values.
+For example, the scheduler gives thread A of priority 4 *higher* priority
+over thread B of priority 7; likewise thread C of priority -2 has higher
+priority than both thread A and thread B.
+
+The scheduler distinguishes between two classes of threads,
+based on each thread's priority.
+
+* A :dfn:`cooperative thread` has a negative priority value.
+  Once it becomes the current thread, a cooperative thread remains
+  the current thread until it performs an action that makes it unready.
+
+  .. image:: cooperative.svg
+     :align: center
+
+* A :dfn:`preemptible thread` has a non-negative priority value.
+  Once it becomes the current thread, a preemptible thread may be supplanted
+  at any time if a cooperative thread, or a preemptible thread of higher
+  or equal priority, becomes ready.
+
+  .. image:: preemptive.svg
+     :align: center
+
+A thread's initial priority value can be altered up or down after the thread
+has been started. Thus it possible for a preemptible thread to become
+a cooperative thread, and vice versa, by changing its priority.
+
+The kernel supports a virtually unlimited number of thread priority levels.
+The configuration options :option:`CONFIG_NUM_COOP_PRIORITIES` and
+:option:`CONFIG_NUM_PREEMPT_PRIORITIES` specify the number of priority
+levels for each class of thread, resulting the following usable priority
+ranges:
+
+* cooperative threads: (-:option:`CONFIG_NUM_COOP_PRIORITIES`) to -1
+* preemptive threads: 0 to (:option:`CONFIG_NUM_PREEMPT_PRIORITIES` - 1)
```

Comment:
Picture says ``CONFIG_NUM_PREEMPT_PRIORITIES + 1`` but the text here says ``CONFIG_NUM_PREEMPT_PRIORITIES- 1``